### PR TITLE
[BOLT] Drop high discrepancy profiles in matching

### PR DESCRIPTION
**Pull request by shawbyoung**
Summary: Functions with high discrepancy 
(measured by matched function blocks) 
can be ignored with an added command line 
argument for better performance.

Test Plan: Added 
stale-matching-min-matched-block.test

### CODE REVIEW

File: bolt/docs/CommandLineArgumentReference.md [hunk 1]
```diff
@@ -614,6 +614,17 @@
...
 - `--lite-threshold-pct=<uint>`
 
+  Threshold (in percent) of matched profile at which stale profile inference is
+  applied to functions. Argument corresponds to the sum of matched execution
+  counts of function blocks divided by the sum of execution counts of function
+  blocks. E.g if the sum of a function blocks' execution counts is 100, the sum
+  of the function blocks' matched execution counts is 10, and the argument is 15
+  (15%), profile inference will not be applied to that function. A higher
+  threshold will correlate with fewer functions to process in cases of stale
+  profile. Default set to %5.
```
Comment (WenleiHe) [inline bolt/docs/CommandLineArgumentReference.md:624]: nit: this is too verbose of a description. as you can see it's longer than most of other descriptions. :)

File: llvm/include/llvm/Transforms/Utils/SampleProfileInference.h [hunk 2]
```diff
@@ -59,6 +59,8 @@ struct FlowFunction {
   /// The index of the entry block.
   uint64_t Entry{0};
   uint64_t Sink{UINT64_MAX};
+  // Matched execution count for the function.
+  uint64_t MatchedExecCount{0};
```
Comment (WenleiHe) [inline llvm/include/llvm/Transforms/Utils/SampleProfileInference.h:63]: nit: I'd be careful about adding this to `FlowFunction` -- strictly speaking this doesn't belong to flow function, which just describe the CFG and if we add function level "attributes" to flow functions, we'd have a lot more here.

File: bolt/lib/Profile/StaleProfileMatching.cpp [hunk 3]
```diff
@@ -592,10 +599,15 @@ void preprocessUnreachableBlocks(FlowFunction &Func) {
 /// Decide if stale profile matching can be applied for a given function.
 /// Currently we skip inference for (very) large instances and for instances
 /// having "unexpected" control flow (e.g., having no sink basic blocks).
-bool canApplyInference(const FlowFunction &Func) {
+bool canApplyInference(const FlowFunction &Func,
```
Comment (WenleiHe) [inline bolt/lib/Profile/StaleProfileMatching.cpp:602]: Header comment needs update.

File: bolt/lib/Profile/StaleProfileMatching.cpp [hunk 4]
```diff
@@ -592,10 +599,15 @@ void preprocessUnreachableBlocks(FlowFunction &Func) {
...
+                       const yaml::bolt::BinaryFunctionProfile &YamlBF) {
   if (Func.Blocks.size() > opts::StaleMatchingMaxFuncSize)
     return false;
 
+  if ((double)Func.MatchedExecCount / YamlBF.ExecCount >=
+      opts::MatchedProfileThreshold / 100.0)
+    return false;
```
Comment (WenleiHe) [inline bolt/lib/Profile/StaleProfileMatching.cpp:609]: Trying to understand the rationale behind using dynamic counts to determine whether profile inference is safe. 

The way I see it is, we have two graph that we try to match, if we have many nodes in the graph that we have exact match, chances are higher that we can infer the correct match for the rest of the nodes. With that, we care about more how many nodes we can match statically. 

Say if we have 5 blocks with count distribution of 1M, 1K, 1K, 1k, 1K, if we have exact match for the 4 1K node (80% exact match), we should feel reasonably confident about inferring the remaining 1 node, even though if we look at counts, we have exact match for only <1%. 

WDYT?

**Comment by WenleiHe** [comment]
cc @wlei-llvm

File: bolt/lib/Profile/StaleProfileMatching.cpp (hunk 4 above)
Comment (aaupov) [reply bolt/lib/Profile/StaleProfileMatching.cpp:609]: It's a tricky question how to define the cutoff in terms of sufficient matching.
I first thought of defining a block count based cutoff (if we matched >5% of blocks, proceed with matching), but then what if these are cold blocks covering <1% of exec count? In this case we'd end up guessing/propagating most samples.

For block-based matching, the threshold should be higher than 5%, perhaps closer to a half? For exec count based matching, I'd feel comfortable with 5% as threshold.

Comment (WenleiHe) [reply bolt/lib/Profile/StaleProfileMatching.cpp:609]: > For block-based matching, the threshold should be higher than 5%, perhaps closer to a half?

Yes. Threshold of course need to be tuned based on the heuristic chosen. I just feel that block count based threshold could be a better proxy of how confident we are about the graph match and whether stale profile matching should proceed.. But I don't have very strong opinion.

Comment (shawbyoung) [reply bolt/lib/Profile/StaleProfileMatching.cpp:609]: I’m leaning towards the block count heuristic now. I think the 1M and 4x1K exec count block case is likely pretty common – I imagine functions with loops would look a lot like this. Having some blocks matched exactly would suggest to me that there would likely be a reasonable amount of similarity between the profiled function and existing function relationally, which block coldness likely doesn’t have an outsized bearing on. I think having a reasonably high threshold for matched blocks would conservatively allow us to drop functions in high discrepancy – I’ll test this on a production binary.

File: bolt/lib/Profile/StaleProfileMatching.cpp [hunk 5]
```diff
@@ -51,6 +51,12 @@ cl::opt<bool>
                       cl::desc("Infer counts from stale profile data."),
                       cl::init(false), cl::Hidden, cl::cat(BoltOptCategory));
 
+cl::opt<unsigned> MatchedProfileThreshold(
+    "matched-profile-threshold",
```
Comment (WenleiHe) [inline bolt/lib/Profile/StaleProfileMatching.cpp:55]: nit: the current name is quite ambiguous, suggest rename it to `stale-matching-min-matched-block` to be consistent with the convention used by other switches, and also more explicit. 

Suggestion for description: "Minimum number of exact match block for a function to be considered for profile inference."

File: llvm/include/llvm/Transforms/Utils/SampleProfileInference.h [hunk 6]
```diff
@@ -58,7 +58,6 @@ struct FlowFunction {
   std::vector<FlowJump> Jumps;
   /// The index of the entry block.
   uint64_t Entry{0};
-  uint64_t Sink{UINT64_MAX};
```
Comment (WenleiHe) [inline llvm/include/llvm/Transforms/Utils/SampleProfileInference.h:61]: this change doesn't belong to this PR

File: bolt/lib/Profile/StaleProfileMatching.cpp [hunk 7]
```diff
@@ -180,6 +186,17 @@ struct BlendedBlockHash {
...
 
+/// A data object containing function matching information.
+struct FunctionMatchingData {
+public:
+  /// The number of blocks matched exactly.
+  uint64_t MatchedExactBlocks{0};
+  /// The number of blocks matched loosely.
+  uint64_t MatchedLooseBlocks{0};
+  /// The number of execution counts matched.
+  uint64_t MatchedExecCounts{0};
```
Comment (WenleiHe) [inline bolt/lib/Profile/StaleProfileMatching.cpp:197]: What's the intended use for these two? There is no usage anywhere right now, and usually we don't keep unused stuff around.

Comment (shawbyoung) [reply bolt/lib/Profile/StaleProfileMatching.cpp]: Left this for possible extensibility of our definition for the threshold, but since these aren't being used at the moment, I'll remove them

File: bolt/lib/Profile/StaleProfileMatching.cpp [hunk 8]
```diff
@@ -51,6 +51,12 @@ cl::opt<bool>
                       cl::desc("Infer counts from stale profile data."),
                       cl::init(false), cl::Hidden, cl::cat(BoltOptCategory));
 
+cl::opt<unsigned> StaleMatchingMinMatchedBlock(
+    "stale-matching-min-matched-block",
+    cl::desc("Percentage threshold of matched basic blocks at which stale "
+             "profile inference is executed."),
+    cl::init(50), cl::Hidden, cl::cat(BoltOptCategory));
```
Comment (WenleiHe) [inline bolt/lib/Profile/StaleProfileMatching.cpp:58]: Do we want to leave the default same as current behavior for now?

Comment (shawbyoung) [reply bolt/lib/Profile/StaleProfileMatching.cpp]: Yes, changed the default to 0

File: bolt/lib/Profile/StaleProfileMatching.cpp [hunk 9]
```diff
@@ -180,6 +186,13 @@ struct BlendedBlockHash {
   uint8_t SuccHash{0};
 };
 
+/// A data object containing function matching information.
+struct FunctionMatchingData {
+public:
```
Comment (WenleiHe) [inline bolt/lib/Profile/StaleProfileMatching.cpp:191]: nit: public is not needed for struct since it's the default. also if we only have one counter, the struct seems like an overkill. we could just keep it simple and use an integer counter for now without a struct, and expand later if needed.

**Review by WenleiHe — APPROVED** [review]
lgtm with a nit, thanks.

File: bolt/lib/Profile/StaleProfileMatching.cpp [hunk 10]
```diff
@@ -725,18 +739,22 @@ bool YAMLProfileReader::inferStaleProfile(
   const BinaryFunction::BasicBlockOrderType BlockOrder(
       BF.getLayout().block_begin(), BF.getLayout().block_end());
 
+  // Tracks the number of matched blocks.
+  uint64_t MatchedBlocks;
```
Comment (WenleiHe) [inline bolt/lib/Profile/StaleProfileMatching.cpp:743]: nit: use explicit initialization to be safe.

File: bolt/lib/Profile/StaleProfileMatching.cpp [hunk 11]
```diff
@@ -575,10 +583,16 @@ void preprocessUnreachableBlocks(FlowFunction &Func) {
...
 /// Currently we skip inference for (very) large instances and for instances
 /// having "unexpected" control flow (e.g., having no sink basic blocks).
-bool canApplyInference(const FlowFunction &Func) {
+bool canApplyInference(const FlowFunction &Func,
+                       const yaml::bolt::BinaryFunctionProfile &YamlBF,
+                       const uint64_t &MatchedBlocks) {
   if (Func.Blocks.size() > opts::StaleMatchingMaxFuncSize)
     return false;
 
+  if ((double)(MatchedBlocks) / YamlBF.Blocks.size() <=
```
Comment (aaupov) [inline bolt/lib/Profile/StaleProfileMatching.cpp:592]: It's preferred to use integer math when dealing with options, e.g.:
https://github.com/llvm/llvm-project/blob/67285feffd6708e6db36c11faf95eeab449e797a/bolt/lib/Passes/IndirectCallPromotion.cpp#L584-L586

File: bolt/lib/Profile/StaleProfileMatching.cpp [hunk 12]
```diff
@@ -394,7 +400,7 @@ createFlowFunction(const BinaryFunction::BasicBlockOrderType &BlockOrder) {
 void matchWeightsByHashes(BinaryContext &BC,
                           const BinaryFunction::BasicBlockOrderType &BlockOrder,
                           const yaml::bolt::BinaryFunctionProfile &YamlBF,
-                          FlowFunction &Func) {
+                          FlowFunction &Func, uint64_t &MatchedBlocksCount) {
```
Comment (aaupov) [inline bolt/lib/Profile/StaleProfileMatching.cpp:403]: Why not return MatchedBlocksCount instead?

Comment (shawbyoung) [reply bolt/lib/Profile/StaleProfileMatching.cpp:403]: In the case that we want to build upon our profile discrepancy heuristic, I think it makes sense to pass in a reference to "something" (variable, struct) that maintains information related to the heuristic, similarly how to the flow function is treated by this function. But I don't have strong feelings about this - if you do I'm happy to change it.

Comment (aaupov) [reply bolt/lib/Profile/StaleProfileMatching.cpp:403]: I think it makes sense to return the stats as a result of matching. We can extend the return type into struct in the future if the need arises.

Comment (shawbyoung) [reply bolt/lib/Profile/StaleProfileMatching.cpp:403]: Sounds good

File: bolt/lib/Profile/StaleProfileMatching.cpp [hunk 13]
```diff
@@ -391,10 +397,9 @@ createFlowFunction(const BinaryFunction::BasicBlockOrderType &BlockOrder) {
 /// of the basic blocks in the binary, the count is "matched" to the block.
 /// Similarly, if both the source and the target of a count in the profile are
 /// matched to a jump in the binary, the count is recorded in CFG.
-void matchWeightsByHashes(BinaryContext &BC,
-                          const BinaryFunction::BasicBlockOrderType &BlockOrder,
-                          const yaml::bolt::BinaryFunctionProfile &YamlBF,
-                          FlowFunction &Func) {
+uint64_t matchWeightsByHashes(
```
Comment (aaupov) [inline bolt/lib/Profile/StaleProfileMatching.cpp:400]:
```suggestion
size_t matchWeightsByHashes(
```

**Review by aaupov — APPROVED** [review]
LGTM with one nit
